### ISSUE TITLE
[PHP8.2] Support for fetching property of enums in const expressions

### ISSUE THREAD

**Issue by Wirone** [PHP-CS-Fixer/PHP-CS-Fixer#6613]
The goal of this task is providing support for [fetching property of enum in const expressions](https://wiki.php.net/rfc/fetch_property_in_const_expressions).

Code that uses syntax like `Enum::Foo->name` or `Enum::Foo->value` must not break any fixers. All tests related to such usage must `@require` PHP 8.2

**TODO**:

- [x] Add [more fixtures](#issuecomment-1241955456) and fix them → #6803

**Comment by SpacePossum** [comment]
Thanks for adding. The syntax itself doesn't bring anything new does? I think it is only allowed on more types, as such I don't think we need to do any functional changes, but good to add tests (at least to the yet to be create PHP8.2 integration test)

**Comment by Wirone** [comment]
@SpacePossum I guess it does bring something new. Example from RFC:

```php
enum A: string {
    case B = 'B';
    // This is currently not permitted
    const C = [self::B->value => self::B];
}
```

Since these are const expressions, `->` is not expected there. Minimum that should be done is ensuring it does not break anything. If code doesn't need to be changed, the better 😉

**Comment by Wirone** [comment]
[As I found out](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/6621#issuecomment-1237500506) in #6621 fetching property of enums in const expressions works properly (surprisingly for me), BUT we probably need fixer that would enforce lowercase `->name` and `->value` for enums, because [any other casing triggers a warning](https://3v4l.org/laSPV/rfc#vgit.master). I don't know if it's must have or only nice to have since such code won't work as expected, so people need to change it on their own anyway. But maybe it's worth doing - let me know what you think 🙂

**Comment by SpacePossum** [comment]
Thanks for all the research!
We still need to check all fixers running on this new PHP8.2 code to be sure, so not just the current utests and integration tests.
For the lower case `->name` and `->value`, I don't think the tool has the scope or knows if `->name` is called on an enum or an object (as we have only file scope, so no AST or project wide analyses of some kind), so good idea but I'm not sure it can be done.

**Comment by Wirone** [comment]
At least we could detect if `->name` / `->value` is after `Foo::Bar` syntax (fetching property in const expressions), because for `$foo->name`, as you pointed, it would not be possible.

How do you see running all fixers? Is there something like that for other PHP versions?

**Comment by SpacePossum** [comment]
>At least we could detect if ->name / ->value is after Foo::Bar syntax (fetching property in const expressions), because for $foo->name, as you pointed, it would not be possible.

:+1: 

>How do you see running all fixers? Is there something like that for other PHP versions?

we have to search the code base for all fixers dealing with this type of syntax. Like all touching `::` and all touching `->` and add tests and see if all goes fine. It is manually thingy which takes some time :sweat_smile:

**Comment by Wirone** [comment]
After #6621 gets merged, we need to create another MR that would add and fix such fixtures:

<details>
  <summary>SHOW DIFF</summary>

```diff
Index: tests/Fixtures/Integration/misc/PHP8_2.test
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/Fixtures/Integration/misc/PHP8_2.test b/tests/Fixtures/Integration/misc/PHP8_2.test
--- a/tests/Fixtures/Integration/misc/PHP8_2.test	(revision 6c6655389b3b2d6d5f0ebaab6248daf10b76e1d7)
+++ b/tests/Fixtures/Integration/misc/PHP8_2.test	(date 1662728694526)
@@ -64,6 +64,41 @@
     private const THREE = 'three';
 }
 
+// https://wiki.php.net/rfc/fetch_property_in_const_expressions
+enum E: string
+{
+    case Foo = 'foo';
+}
+
+$eFoo = E::Foo->value;
+
+function e1()
+{
+    static $v = E::Foo->value;
+}
+
+#[Attr(E::Foo->name)]
+class E2
+{
+}
+
+function e3(
+    $p = E::Foo->value,
+) {}
+
+class E4
+{
+    public string $p = E::Foo->name;
+}
+
+const E5 = 'value';
+class E6
+{
+    public const C = E::Foo->{E5};
+}
+
+array_filter([E::Foo->name], fn (string $a) => $a === E::Foo->value);
+
 --INPUT--
 <?php
 
@@ -117,3 +152,34 @@
     protected const TWO   = 'two';
     private   const THREE = 'three';
 }
+
+// https://wiki.php.net/rfc/fetch_property_in_const_expressions
+enum E: string {
+    case   Foo   =   'foo';
+}
+
+$eFoo = E::Foo -> value;
+
+function e1() {
+    static $v = E::Foo -> VALUE;
+}
+
+#[Attr(E::Foo -> NAME)]
+class E2
+{
+}
+
+function e3(
+    $p = E::Foo -> Value,
+) {}
+
+class E4 {
+    public     string $p = E::Foo->Name;
+}
+
+const E5 = 'value';
+class E6 {
+    public     const C   =   E::Foo -> {E5};
+}
+
+array_filter([ E::Foo->NAME ], fn   (string   $a)   =>   $a   ===   E::Foo->VALUE);

```
</details>

Those fixture were failing, so they were removed in order to be able to merge integration test's base.

**Comment by Wirone** [comment]
In #6803 regression tests were added and it seems like fetching property of enums in const expressions are supported properly. Fixtures described [above](#issuecomment-1241955456) are not needed because only `->name` and `->value` are valid fetches, other cases would be marked by IDE as invalid and won't work:

![image](https://user-images.githubusercontent.com/600668/223761654-10ee5a74-68cc-41a6-94c7-f42cff07f751.png)

![image](https://user-images.githubusercontent.com/600668/223761886-9f0582d7-0f32-4e46-b461-352915adbf61.png)

Fixer for such fetches is of course possible, but totally as nice to have. I'm going to close this issue now as completed 🙂
